### PR TITLE
cmd/operator-sdk/internal/genutil: relax version dir check in `parseGroupVersions()`

### PR DESCRIPTION
**Pull request by estroz**
**Description of the change:** remove requirement that the API dir structure `pkg/apis/<group>/<version>` have `<version>` match something like `v1alpha1`, and instead check if `<version>` is a dir (or dir hierarchy) that contains files.

**Motivation for the change:** having a API dir shared by multiple versions, ex `pkg/apis/operators/shared`, is valid and should be considered when running a code generator.

### CODE REVIEW

File: cmd/operator-sdk/internal/genutil/genutil.go [hunk 1]
```diff
@@ -45,8 +45,20 @@ func parseGroupVersions() (map[string][]string, error) {
...
+				if v.IsDir() {
+					// Ignore directories that do not contain any files, so generators
+					// do not get empty directories as arguments.
+					verDir := filepath.Join(groupDir, v.Name())
+					files, err := ioutil.ReadDir(verDir)
+					if err != nil {
+						return nil, fmt.Errorf("could not read %s directory to find api Versions: %v", verDir, err)
+					}
+					for _, f := range files {
+						if !f.IsDir() {
```
Comment (joelanford) [inline cmd/operator-sdk/internal/genutil/genutil.go]: Should we also check to see if the file has `.go` extension, or is the code generation okay with directories containing _any_ files?

Comment (estroz) [reply cmd/operator-sdk/internal/genutil/genutil.go]: We should. I'm not sure if the generator would error in that case, but its better to check anyway.

**Comment by joelanford** [comment]
Before merging, we should double check with @awgreene that this solves his problem.

**Comment by awgreene** [comment]
> Before merging, we should double check with @awgreene that this solves his problem.

Thanks for working on this @joelanford and @estroz!

I pulled the `relax-fqapis-check` branch and ran the sdk after updating the phase types to include the `deepcopy gen` code:
```
// Phase describes the phase the object is in
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
type Phase struct {
	// Name of the phase
	Name string `json:"name,omitempty"`

	// A human readable message indicating why the object is in this phase
	Message string `json:"message,omitempty"`
}

// ObjectPhase describes the phase of a Marketplace object is in along with the
// last time a phase transition occurred and when the object was last updated
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
type ObjectPhase struct {
	// Current phase of the object
	Phase `json:"phase,omitempty"`

	// Last time the object has transitioned from one phase to another
	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`

	// Last time the status of the object was updated
	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty"`
}
```

I then ran the following command:
```
$ operator-sdk generate k8s
INFO[0000] Running deepcopy code-generation for Custom Resource group versions: [operators:[shared v1 v2], ] 
INFO[0001] Code-generation complete.                           
```

It looks like the shared dir was inspected, but there was no code generated. Any thoughts?

**Comment by estroz** [comment]
@awgreene do you have `doc.go` and `register.go` in `pkg/apis/operator/shared`?

Also you might have to have `Phase` look like the following:

```Go
type Phase struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	...
}
```

While testing with your code I got compilation errors because `Phase` didn't implement the `k8s.io/apimachinery/pkg/runtime.Object` method, which requires a `GetObjectKind()` method.

**Comment by awgreene** [comment]
> ```go
> type Phase struct {
> 	metav1.TypeMeta   `json:",inline"`
> 	metav1.ObjectMeta `json:"metadata,omitempty"`
> ```

Thanks for looking into this @estroz, is there a reason why this needs to be added if it's moved into it's own package? When `phase` was in the `v1` package it did not need the field.

**Comment by estroz** [comment]
@awgreene can you link the rest of your code?

**Comment by awgreene** [comment]
> @awgreene can you link the rest of your code?

https://github.com/awgreene/operator-marketplace/commit/7d6b64f9dc094661e54dcca3f6b5bb68a183410f

I copied the old deepcopy code into this commit earlier but the only change other change involved moving the `phase_types.go` file.

**Comment by estroz** [comment]
Ah I see the issue. My test code had `// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object` above `ObjectPhase` and `Phase` which forces `DeepCopyObject()` to be generated, but that is not necessary. Disregard my field addition comment.

**Comment by awgreene** [comment]
> Ah I see the issue. My test code had `// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object` above `ObjectPhase` and `Phase` which forces `DeepCopyObject()` to be generated, but that is not necessary. Disregard my field addition comment.

Great - adding the doc file seems to have fixed everything! Thanks @estroz

**Pull request merged**